### PR TITLE
InstData Import: Merge and Modify Sanitize and Split Bam

### PR DESCRIPTION
**Pull request by ebelter**
merge sanitize and split commadns
split reads into paired and singletons (instead of read1 & read2)
reads get consistent flags based on if paired/singleton
collect read 1/2s including alternate alignments, keeping one from each
discard supplemental reads

### CODE REVIEW

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+    ],
+    has_optional_transient => [
+        old_and_new_read_group_ids => { is => 'HASH', default => {}, },
+        _read_group_fhs => { is => 'HASH', default => {} },
+    ],
+};
+
+sub execute {
+    my $self = shift;
+    $self->debug_message('Spilt bam by read group...');
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm]: These entry and exit debug messages could be updated to reflect the widened responsibility of this class.

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+
+sub _process_reads {
+    my ($self) = @_;
+    $self->debug_message('Processing reads...');
+
+    my $bam_path = $self->bam_path;
+    $self->debug_message("Bam path: $bam_path");
+    my $bam_fh = IO::File->new("samtools view $bam_path |");
+    if ( not $bam_fh ) {
+        die $self->error_message('Failed to open file handle to samtools command!');
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm]: Obligatory plug for `$self->fatal_message`!

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+    $type;
+}
+
+sub _sanitize_read {
+    # Upcase sequence
+    $_[0]->[9] = uc $_[0]->[9];
+    # If listed as revcomp, revcomp seq and qual
+    if ( $_[0]->[1] & 0x10 ) {
+        $_[0]->[9] = reverse $_[0]->[9];
+        $_[0]->[9] =~ tr/ATCG/TAGC/;
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm]: It seems like we ought to have a central reverse complement implementation somewhere!  A `grep` of the code finds we have a few different implementations around, though none of them are in a convenient spot.

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    my ($self, $read_group_id, $type) = @_;
+
+    my $key = join('*', $read_group_id, $type);
+    unless ( exists $self->_read_group_fhs->{$key} ) {
+        my $read_group_bam_path = $self->get_working_bam_path_with_new_extension($self->bam_path, $read_group_id, $type);
+        my $samtools_cmd = "| samtools view -S -b -o $read_group_bam_path -";
+        $self->debug_message("Opening fh for $read_group_bam_path $type with:\n$samtools_cmd");
+        my $fh = IO::File->new($samtools_cmd);
+        if ( not $fh ) {
+            $self->fata_message('Failed to open file handle to samtools command!');
```
Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm]: `fatal_message` is short of an `l`.

File: lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm (hunk 3 above)
Comment (ebelter) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm]: Agreed! Alas, it is for another time.

Comment (tmooney) [inline lib/perl/Genome/InstrumentData/Command/Import/WorkFlow/SanitizeAndSplitBam.pm]: Alas and alack!

**Comment by tmooney** [comment]
+1 :boar:

**Pull request merged**